### PR TITLE
Event-Filter Batch 4 - Extended Operators

### PR DESCRIPTION
**Pull request by andreasebner**
This PR is the fourth batch of the event filter server logic.

This batch contains:

-> Dedicated Unit-Test for event filters
-> isNull Operator
-> Type conversion rules and compare operator
-> bitwise operator
-> between operator
-> inList operator

The PR is based on the work in #4225

### CODE REVIEW

File: src/server/ua_subscription_events.c [hunk 1]
```diff
@@ -409,6 +409,593 @@ orOperator(UA_Server *server, UA_Session *session,
...
+        secondCompareOperand = firstOperand;
+    }
+    UA_SByte castRule =
+        checkTypeCastingOption(firstCompareOperand->type, secondCompareOperand->type);
+
+    if(!(castRule == 0 || castRule == 1)){
+        return UA_STATUSCODE_BADTYPEMISMATCH;
+    }
+
+    UA_Int32 rule = -1;
```
Comment (jpfr) [inline src/server/ua_subscription_events.c:582]: This is easy to misunderstand.
The castRule is -1/0/1/2. This rule here has many more possibilities.
Express the difference in the wording. And/or make it an enum. And add a comment.

File: src/server/ua_subscription_events.c [hunk 2]
```diff
@@ -409,6 +409,593 @@ orOperator(UA_Server *server, UA_Session *session,
...
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_XMLELEMENT){
+        rule = 12;
+    } else if(castRule == 1 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_BOOLEAN &&
+              isNumericUnsigned(secondCompareOperand->type->typeKind)){
+        rule = 100;
+    } else if(castRule == 1 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_BOOLEAN &&
+              isNumericSigned(secondCompareOperand->type->typeKind)) {
+        rule = 101;
```
Comment (jpfr) [inline src/server/ua_subscription_events.c:629]: Eine ganze Reihe dieser Rules werden weiter unten nicht verwendet.
Warum weisen wir dann hier 101 zu?

File: src/server/ua_subscription_events.c [hunk 3]
```diff
@@ -409,6 +409,593 @@ orOperator(UA_Server *server, UA_Session *session,
...
+
+    UA_Byte firstOperatorTypeKindIndex = UA_BYTE_MAX, secondOperatorTypeKindIndex = UA_BYTE_MAX;
+    for(UA_Byte i = 0; i < 21; ++i) {
+        if(cast_target->typeKind == kindConversionRule[i])
+            firstOperatorTypeKindIndex = i;
+        if(cast_source->typeKind == kindConversionRule[i])
+            secondOperatorTypeKindIndex = i;
+    }
+
+    //0 -> Same Type, 1 -> Implicit Cast, 2 -> Only explicit Cast, -1 -> cast invalid
```
Comment (jpfr) [inline src/server/ua_subscription_events.c:519]: The explicit cast is never used.
Is that "lossy" casting?

File: src/server/ua_subscription_events.c [hunk 4]
```diff
@@ -409,6 +409,593 @@ orOperator(UA_Server *server, UA_Session *session,
...
+                return UA_STATUSCODE_GOOD;
+            break;}
+        case 2:{
+            UA_Int64 int64_o1 = 0, int64_o2 = 0;
+            implicitNumericCast(firstCompareOperand, &UA_TYPES[UA_TYPES_INT64], &int64_o1);
+            implicitNumericCast(secondCompareOperand, &UA_TYPES[UA_TYPES_INT64], &int64_o2);
+            if(int64_o1 == int64_o2)
+                return UA_STATUSCODE_GOOD;
+            break;}
+        case 3:
```
Comment (jpfr) [inline src/server/ua_subscription_events.c:651]: Check the definition of `UA_order`.
With that you can cover most of these comparison operations with a single code-path.
Furthermore we can also check the equality of structure-typed values.

**Review by jpfr — COMMENTED** [review]
Looks good!
See the detailed comments.

File: src/server/ua_subscription_events.c [hunk 5]
```diff
@@ -409,6 +409,593 @@ orOperator(UA_Server *server, UA_Session *session,
...
+        UA_DATATYPEKIND_LOCALIZEDTEXT,
+        UA_DATATYPEKIND_QUALIFIEDNAME,
+        UA_DATATYPEKIND_UINT16,
+        UA_DATATYPEKIND_UINT32,
+        UA_DATATYPEKIND_UINT64,
+        UA_DATATYPEKIND_XMLELEMENT
+    };
+
+    UA_Byte firstOperatorTypeKindIndex = UA_BYTE_MAX, secondOperatorTypeKindIndex = UA_BYTE_MAX;
+    for(UA_Byte i = 0; i < 21; ++i) {
```
[the remark this answers is not in the document]
Comment (andreasebner) [reply src/server/ua_subscription_events.c:512]: The special order within the array is relevant. Therefore, it's not enough to loop over the standard UA_DATATYPEKIND elements.

[the remark this answers is not in the document]
Comment (jpfr) [reply src/server/ua_subscription_events.c:512]: You could have a 30-element array that is a direct map to the target index.
No need to loop to find the new index.

**Pull request merged**